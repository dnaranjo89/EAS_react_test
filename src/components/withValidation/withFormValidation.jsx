import React, { Component } from 'react';
import PropTypes from 'prop-types';

const withFormValidation = WrappedComponent => {
  class WithFormValidation extends Component {
    constructor(props) {
      super(props);
      this.state = {
        formSubmitted: false,
        validations: {},
        formError: 'the is an error',
        changedFields: [],
      };
    }

    getChildContext() {
      return {
        registerValidatedField: this.registerValidatedField.bind(this),
        deregisterValidatedField: this.deregisterValidatedField.bind(this),
        onFieldChange: this.onFieldChange.bind(this),
        getFieldErrors: this.getFieldErrors.bind(this),
        getFormError: this.getFormError.bind(this),
        updateErrors: this.updateErrors.bind(this),
        updateFieldChangedState: this.updateFieldChangedState.bind(this),
      };
    }

    componentDidUpdate = (previousProps, previousState) => {
      // Check global form errors
      const formError = this.props.checkErrors();
      if (previousState.formError !== formError) {
        this.setState({
          formError,
        });
      }
    };

    onSubmit = e => {
      if (!this.isFormValid()) {
        e.preventDefault();
        this.setState({
          formSubmitted: true,
        });
      } else {
        this.props.onSubmit(e);
      }
    };

    onFieldChange = (name, valid) => {
      // TODO not sure why this is needed, remove if after a while doesn't cause problems
      // this.updateErrors(name, valid);
      // this.updateFieldChangedState(name);
    };

    getFieldErrors(name) {
      const { changedFields, validations, formSubmitted } = this.state;
      // return formSubmitted || changedFields.includes(name) ? validations[name] : undefined;  //don't rememeber why formSubmitted was needed here
      return changedFields.includes(name) ? validations[name] : undefined;
    }

    getFormError() {
      return this.state.formError;
    }

    isFormValid = () => {
      const validations = Object.values(this.state.validations);
      const fieldsValid = !validations.some(Boolean);
      const formValid = !this.state.formError;
      return fieldsValid && formValid;
    };

    registerValidatedField(name, valid, value) {
      this.updateErrors(name, valid);

      const isEmptyAtRegister = (Array.isArray(value) && !value.length) || value === '';
      if (!isEmptyAtRegister) {
        this.updateFieldChangedState(name);
      } else if (this.state.changedFields.indexOf(name) !== -1) {
        const changedFields = this.state.changedFields.slice();
        changedFields.splice(changedFields.indexOf(name));
        this.setState({
          changedFields,
        });
      }
    }

    updateFieldChangedState(name) {
      if (!this.state.changedFields.includes(name)) {
        const changedFields = this.state.changedFields.slice();
        changedFields.push(name);
        this.setState(() => ({
          changedFields,
        }));
      }
    }

    updateErrors(name, errors) {
      const formError = this.props.checkErrors();
      this.setState(
        previousState => ({
          formError,
          validations: {
            ...previousState.validations,
            ...{
              [name]: errors,
            },
          },
        }),
        () => {
          this.triggerValidationChange();
        },
      );
    }

    triggerValidationChange = () => {
      if (this.props.onValidationChange) {
        const isValid = this.isFormValid();
        this.props.onValidationChange(isValid);
      }
    };

    deregisterValidatedField(name) {
      this.setState(
        previousState => {
          const validations = Object.assign({}, previousState.validations);
          delete validations[name];
          return {
            validations,
          };
        },
        () => {
          if (this.props.onFieldDeregister) {
            this.props.onFieldDeregister(name);
          }
          this.triggerValidationChange();
        },
      );
    }

    render() {
      const { onSubmit, onFieldDeregister, onValidationChange, ...props } = this.props;
      return <WrappedComponent {...props} onSubmit={this.onSubmit} />;
    }
  }

  WithFormValidation.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    onFieldDeregister: PropTypes.func,
    onValidationChange: PropTypes.func,
    checkErrors: PropTypes.func,
  };

  WithFormValidation.defaultProps = {
    onFieldDeregister: null,
    onValidationChange: null,
    checkErrors: () => {},
  };

  WithFormValidation.childContextTypes = {
    registerValidatedField: PropTypes.func,
    deregisterValidatedField: PropTypes.func,
    onFieldChange: PropTypes.func,
    getFieldErrors: PropTypes.func,
    getFormError: PropTypes.func,
    updateErrors: PropTypes.func,
    updateFieldChangedState: PropTypes.func,
  };

  return WithFormValidation;
};

export default withFormValidation;
