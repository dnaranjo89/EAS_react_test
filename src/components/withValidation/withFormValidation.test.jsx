import React from 'react';
import PropTypes from 'prop-types';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import withFormValidation from './withFormValidation';

jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the translate HoC receive the t function as a prop
  translate: () => Component => {
    // eslint-disable-next-line no-param-reassign
    Component.defaultProps = { ...Component.defaultProps, t: () => 'some translated stuff' };
    return Component;
  },
}));

// eslint-disable-next-line
const Form = ({ onFieldRegister, onFieldDeregister, onFieldChange, ...props }) => (
  <form {...props} />
);

const ValidationFeedbackMock = (props, context) => <div>{context.getFormError()}</div>;

ValidationFeedbackMock.contextTypes = {
  getFormError: PropTypes.func.isRequired,
};

// eslint-disable-next-line
const Field = ({ valid, FormHelperTextProps, helperText, error, ...props }) => <input type="text" {...props} />;

const FormWithValidation = withFormValidation(Form);

describe('withFormValidation', () => {
  it('should render correctly', () => {
    const wrapper = mount(<FormWithValidation onSubmit={jest.fn()} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('Registering fields', () => {
    it('should register validated fields', () => {
      const wrapper = shallow(<FormWithValidation onSubmit={jest.fn()} />);
      const instance = wrapper.instance();

      instance.registerValidatedField('field1', true);
      instance.registerValidatedField('field2', false);

      wrapper.update();

      const { state } = instance;

      expect('field1' in state.validations).toBe(true);
      expect('field2' in state.validations).toBe(true);
      expect(state.validations.field1).toBe(true);
      expect(state.validations.field2).toBe(false);
    });

    it('should deregister validated fields', () => {
      const wrapper = shallow(<FormWithValidation onSubmit={jest.fn()} />);
      const instance = wrapper.instance();

      instance.registerValidatedField('field1', false);
      instance.registerValidatedField('field2', true);

      instance.deregisterValidatedField('field2');
      wrapper.update();

      const { state } = instance;

      expect('field1' in state.validations).toBe(true);
      expect('field2' in state.validations).toBe(false);
    });
  });

  describe('Form errors', () => {
    it('are not shown when there are not', () => {
      const wrapper = mount(
        <FormWithValidation onSubmit={jest.fn()} checkErrors={() => undefined}>
          <ValidationFeedbackMock />
        </FormWithValidation>,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('are not shown if the form is not submitted', () => {
      const wrapper = mount(
        <FormWithValidation onSubmit={jest.fn()} checkErrors={() => 'There are errors'}>
          <ValidationFeedbackMock />
        </FormWithValidation>,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('are not shown when fields change', () => {
      const wrapper = mount(
        <FormWithValidation onSubmit={jest.fn()} checkErrors={() => 'There are errors'}>
          <ValidationFeedbackMock />
        </FormWithValidation>,
      );
      const instance = wrapper.instance();
      instance.updateFieldValidationState('afield', false);
      wrapper.update();
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('are shown when the form is submitted', () => {
      const wrapper = mount(
        <FormWithValidation onSubmit={jest.fn()} checkErrors={() => 'There are errors'}>
          <ValidationFeedbackMock />
        </FormWithValidation>,
      );
      const instance = wrapper.instance();
      instance.setState({ formSubmitted: true });
      wrapper.update();
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
describe('Submit', () => {
  it('Should store submitted state when submitted', () => {
    const wrapper = mount(<FormWithValidation onSubmit={jest.fn()} />);
    const instance = wrapper.instance();
    expect(instance.state.formSubmitted).toBe(false);
    wrapper.find('form').simulate('submit');
    expect(instance.state.formSubmitted).toBe(true);
  });

  it('Should call the main onSubmit function if the form is valid', () => {
    const onSubmitMock = jest.fn();
    const wrapper = mount(<FormWithValidation onSubmit={onSubmitMock} />);

    const instance = wrapper.instance();
    instance.isFormValid = jest.fn(() => true);

    wrapper.find('form').simulate('submit');
    expect(onSubmitMock).toHaveBeenCalled();
  });

  it('Should not call the main onSubmit function if the form is invalid', () => {
    const onSubmitMock = jest.fn();
    const wrapper = mount(<FormWithValidation onSubmit={onSubmitMock} />);

    const instance = wrapper.instance();
    instance.isFormValid = jest.fn(() => false);

    wrapper.find('form').simulate('submit');
    expect(onSubmitMock).not.toHaveBeenCalled();
  });
});
