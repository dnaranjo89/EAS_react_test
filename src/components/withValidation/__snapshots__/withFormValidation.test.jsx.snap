// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFormValidation Form errors are not shown if the form is not submitted 1`] = `
<WithFormValidation
  checkErrors={[Function]}
  onFieldDeregister={null}
  onSubmit={[MockFunction]}
  onValidationChange={null}
>
  <Form
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
    >
      <ValidationFeedbackMock>
        <div />
      </ValidationFeedbackMock>
    </form>
  </Form>
</WithFormValidation>
`;

exports[`withFormValidation Form errors are not shown when fields change 1`] = `
<WithFormValidation
  checkErrors={[Function]}
  onFieldDeregister={null}
  onSubmit={[MockFunction]}
  onValidationChange={null}
>
  <Form
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
    >
      <ValidationFeedbackMock>
        <div />
      </ValidationFeedbackMock>
    </form>
  </Form>
</WithFormValidation>
`;

exports[`withFormValidation Form errors are not shown when there are not 1`] = `
<WithFormValidation
  checkErrors={[Function]}
  onFieldDeregister={null}
  onSubmit={[MockFunction]}
  onValidationChange={null}
>
  <Form
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
    >
      <ValidationFeedbackMock>
        <div />
      </ValidationFeedbackMock>
    </form>
  </Form>
</WithFormValidation>
`;

exports[`withFormValidation Form errors are shown when the form is submitted 1`] = `
<WithFormValidation
  checkErrors={[Function]}
  onFieldDeregister={null}
  onSubmit={[MockFunction]}
  onValidationChange={null}
>
  <Form
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
    >
      <ValidationFeedbackMock>
        <div>
          There are errors
        </div>
      </ValidationFeedbackMock>
    </form>
  </Form>
</WithFormValidation>
`;

exports[`withFormValidation should render correctly 1`] = `
<WithFormValidation
  checkErrors={[Function]}
  onFieldDeregister={null}
  onSubmit={[MockFunction]}
  onValidationChange={null}
>
  <Form
    onSubmit={[Function]}
  >
    <form
      onSubmit={[Function]}
    />
  </Form>
</WithFormValidation>
`;
